File: backend/src/main/java/format/backend/form/dto/FormDetailResponseDto.java
package format.backend.form.dto;

import com.fasterxml.jackson.annotation.JsonFormat;
import format.backend.form.entity.FormShuffleVariant;
import format.backend.form.entity.FormStatus;
import format.backend.form.entity.Language;
import java.time.Duration;
import java.time.Instant;
import java.util.List;
import org.jspecify.annotations.NonNull;
import org.jspecify.annotations.Nullable;

public record FormDetailResponseDto(
        @NonNull String id,

        @NonNull String name,

        @NonNull String slug,

        @Nullable String description,

        @NonNull Language language,

        @NonNull FormStatus status,

        @Nullable FormShuffleVariant shuffleVariant,

        @Nullable String thanksMessage,

        @NonNull Duration estimatedDuration,

        @Nullable String thumbnail,

        @NonNull Boolean allowsQuestionsPreview,

        @NonNull Boolean allowsGuestSubmissions,

        @NonNull Boolean saveSubmissions,

        @NonNull Boolean showAnswersFeedback,

        @NonNull List<@NonNull QuestionResponseDto> questions,

        @NonNull Long submissionsCount,

        @NonNull Long ratingsCount,

        @JsonFormat(shape = JsonFormat.Shape.NUMBER_FLOAT, pattern = "0.0") @Nullable Double ratingAvg,

        @Nullable Integer userRating,

        @Nullable String authorName,

        @NonNull Instant createdAt,

        @NonNull Instant updatedAt) {}

File: frontend/src/core/types/form-detail-response-dto.ts
export interface FormDetailResponseDto {
  readonly id: string;
  readonly name: string;
  readonly description: string | null;
  readonly language: Language;
  readonly formStatus: FormStatus;
  readonly shuffleVariant: FormShuffleVariant | null;
  readonly thanksMessage: string | null;
  readonly estimatedDuration: string;
  readonly thumbnail: string | null;
  readonly allowsQuestionsPreview: boolean;
  readonly allowsGuestSubmissions: boolean;
  readonly saveSubmissions: boolean;
  readonly authorName: string | null;
  readonly submissionsCount: number;
  readonly questions: QuestionResponseDto[];
  readonly createdAt: string;
  readonly updatedAt: string;
}

export enum Language {
  En = "EN",
  Pl = "PL",
}

export enum FormStatus {
  Draft = "DRAFT",
  Public = "PUBLIC",
  Unpublic = "UNPUBLIC",
  Private = "PRIVATE",
  Closed = "CLOSED",
}

export enum FormShuffleVariant {
  Questions = "QUESTIONS",
  Answers = "ANSWERS",
  All = "ALL",
}

export interface QuestionResponseDto {
  readonly id: string;
  readonly content: string;
  readonly type: QuestionType;
  readonly image: string | null;
  readonly isRequired: boolean;
  readonly answers: AnswerResponseDto[];
}

export enum QuestionType {
  SingleChoice = "SINGLE_CHOICE",
  MultipleChoice = "MULTIPLE_CHOICE",
  Open = "OPEN",
}

export interface AnswerResponseDto {
  readonly id: string;
  readonly content: string;
  readonly isCorrect: boolean;
}
